core/trino-main/src/main/java/io/trino/execution/SqlTaskExecution.java

```diff
@@ -587,10 +620,33 @@ public DriverSplitRunner createDriverRunner(@Nullable ScheduledSplit partitioned
             return new DriverSplitRunner(this, driverContext, partitionedSplit);
         }
 
+        /**
+         * @return the created {@link Driver}, or <code>null</code> if the driver factory is already closed because the task is terminating
+         */
+        @Nullable
         public Driver createDriver(DriverContext driverContext, @Nullable ScheduledSplit partitionedSplit)
         {
-            Driver driver = driverFactory.createDriver(driverContext);
+            // Attempt to increment the driver count eagerly, but skip driver creation if the task is already terminating or done
+            if (!driverAndTaskTerminationTracker.tryCreateNewDriver()) {
+                return null;
+            }
+            Driver driver;
+            try {
+                driver = driverFactory.createDriver(driverContext);
+            }
+            catch (Exception e) {
```

Comment:
True, will update this to catch `Throwable` instead but only suppress the exception and return null as long as the driver factory is closed and the caught exception is an `instanceof Exception` (to avoid suppressing fatal errors).